discovery-provider/src/eth_indexing/event_scanner.py

```diff
@@ -0,0 +1,434 @@
+                from_block=_start_block,
+                to_block=_end_block,
+            )
+
+        # Do `n` retries on `eth_get_logs`,
+        # throttle down block range if needed
+        end_block, events = _retry_web3_call(
+            _fetch_events, start_block=start_block, end_block=end_block
+        )
+
+        for evt in events:
+            idx = evt[
+                "logIndex"
+            ]  # Integer of the log index position in the block, null when its pending
+
+            # We cannot avoid minor chain reorganisations, but
+            # at least we must avoid blocks that are not mined yet
+            assert idx is not None, "Somehow tried to scan a pending block"
+
+            block_number = evt["blockNumber"]
+
+            # Get UTC time when this event happened (block mined timestamp)
+            # from our in-memory cache
+            block_when = get_block_when(block_number)
+
+            logger.debug(f'Processing event {evt["event"]}, block:{evt["blockNumber"]}')
+            processed = self.process_event(block_when, evt)
+            all_processed.append(processed)
+
+        return end_block, all_processed
+
+    def estimate_next_chunk_size(self, current_chuck_size: int, event_found_count: int):
+        """Try to figure out optimal chunk size
+
+        Our scanner might need to scan the whole blockchain for all events
+
+        * We want to minimize API calls over empty blocks
+        * We want to make sure that one scan chunk does not try to process too many entries once, as we try to control commit buffer size and potentially asynchronous busy loop
+        * Do not overload node serving JSON-RPC API by asking data for too many events at a time
+
+        Currently Ethereum JSON-API does not have an API to tell when a first event occured in a blockchain
+        and our heuristics try to accelerate block fetching (chunk size) until we see the first event.
+
+        These heurestics exponentially increase the scan chunk size depending on if we are seeing events or not.
+        When any transfers are encountered, we are back to scanning only a few blocks at a time.
+        It does not make sense to do a full chain scan starting from block 1, doing one JSON-RPC call per 20 blocks.
+        """
+
+        if event_found_count > 0:
+            # When we encounter first events, reset the chunk size window
+            current_chuck_size = MIN_SCAN_CHUNK_SIZE
+        else:
+            current_chuck_size *= CHUNK_SIZE_INCREASE
+
+        current_chuck_size = max(MIN_SCAN_CHUNK_SIZE, current_chuck_size)
+        current_chuck_size = min(MAX_CHUNK_SCAN_SIZE, current_chuck_size)
+        return int(current_chuck_size)
+
+    def scan(
+        self,
+        start_block,
+        end_block,
+        start_chunk_size=START_CHUNK_SIZE,
+    ) -> Tuple[list, int]:
+        """Perform a token balances scan.
+
+        Assumes all balances in the database are valid before start_block (no forks sneaked in).
```

Comment:
i dont know if we care about forks necessarily b/c we're not truly "indexing" so this comment might not be necessary